File: frontend/src/api/goals.ts
import { apiClient } from "./client";
import type { EncryptedBlob } from "../crypto/clientEncryption";

export interface GoalSaveResult {
  id: string;
  created_at: string;
}

/** Persists a new goal — ciphertext in, receipt out. No duplicate check:
 * unlike payslip/statement saves, there's no plaintext field the server
 * could dedup on (see backend/db/models.py's Goal docstring), so every
 * call here just creates a new row. */
export async function createGoal(blob: EncryptedBlob): Promise<GoalSaveResult> {
  const { data } = await apiClient.post<GoalSaveResult>("/goals", {
    ciphertext_b64: blob.ciphertextB64,
    iv_b64: blob.ivB64,
  });
  return data;
}

/** Full replace of an existing goal (e.g. after editing savedAmount) —
 * re-encrypt the whole object client-side and send it here. */
export async function updateGoal(id: string, blob: EncryptedBlob): Promise<GoalSaveResult> {
  const { data } = await apiClient.put<GoalSaveResult>(`/goals/${id}`, {
    ciphertext_b64: blob.ciphertextB64,
    iv_b64: blob.ivB64,
  });
  return data;
}

export async function deleteGoal(id: string): Promise<void> {
  await apiClient.delete(`/goals/${id}`);
}

export interface GoalRow {
  id: string;
  ciphertext_b64: string;
  iv_b64: string;
  created_at: string;
}

/** Still ciphertext at this point — decrypt with crypto/clientEncryption.ts before use. */
export async function fetchGoals(): Promise<GoalRow[]> {
  const { data } = await apiClient.get<GoalRow[]>("/goals");
  return data;
}

export interface GoalValuationEntry {
  goal_id: string;
  instrument_type: "fd" | "mutual_fund";
  fd_principal?: number;
  fd_annual_rate?: number;
  fd_start_date?: string;
  mf_scheme_code?: string;
  mf_units_held?: number;
}

export interface GoalValuationResult {
  goal_id: string;
  current_value: number | null;
  error: string | null;
}

/**
 * Live current-value lookup for FD/mutual-fund-linked goals
 * (backend/analytics/investment_valuation.py) — stateless, nothing
 * persisted, same contract as parseStatementText. Called once per Goals
 * tab load (GoalList.tsx), not on every render — an FD's value barely
 * moves minute to minute and a mutual fund's NAV updates once a day, so
 * there's no real value in polling more often than "the tab was opened."
 */
export async function fetchGoalValuations(
  entries: GoalValuationEntry[]
): Promise<GoalValuationResult[]> {
  if (entries.length === 0) return [];
  const { data } = await apiClient.post<GoalValuationResult[]>("/goals/valuation", entries);
  return data;
}

File: backend/api/models/goals.py
"""
Request/response schemas for goal persistence (Goal Tracker UI, V2).
`ciphertext_b64`/`iv_b64` are the AES-256-GCM blob + nonce produced
client-side (§4), base64-encoded for JSON transport — same contract as
api/models/payslip.py, never a plaintext goal figure.

Plaintext shape (once decrypted client-side), for reference — the server
never sees this shape:
    {
      "name": string,               # e.g. "Goa Trip", "Home Loan Down Payment"
      "category": string,           # "Trip" | "Home" | "Education" | "Emergency Fund" | "Retirement" | "Other"
      "targetAmount": number,
      "targetDate": string | null,  # "YYYY-MM-DD", optional
      "savedAmount": number,        # progress so far — the field PUT /goals/{id} updates most often
      "instrumentType": string | null,  # "fd" | "mutual_fund" | absent/null for a plain manual-entry goal (V2.1)
      "fdPrincipal": number | null,      # instrumentType "fd" only
      "fdAnnualRate": number | null,     # instrumentType "fd" only — annual %, e.g. 7.0 for 7%
      "fdStartDate": string | null,      # instrumentType "fd" only — "YYYY-MM-DD"
      "mfSchemeCode": string | null,     # instrumentType "mutual_fund" only — an AMFI scheme code
      "mfUnitsHeld": number | null,      # instrumentType "mutual_fund" only
    }
"""

from pydantic import BaseModel


class GoalSaveRequest(BaseModel):
    ciphertext_b64: str
    iv_b64: str


class GoalOut(BaseModel):
    id: str
    created_at: str


class GoalFull(BaseModel):
    """GET /goals — unlike GoalOut (returned from POST/PUT, just a
    receipt), this carries the ciphertext back so the client can decrypt
    and render every saved goal."""

    id: str
    ciphertext_b64: str
    iv_b64: str
    created_at: str


class GoalValuationRequest(BaseModel):
    """One entry per investment-linked goal to (re-)value — see
    analytics/investment_valuation.py. `goal_id` here is just a caller-
    supplied label to match a response entry back to its goal (the
    client's own local goal id works fine); the server never persists or
    otherwise cares about it."""

    goal_id: str
    instrument_type: str  # "fd" | "mutual_fund"
    fd_principal: float | None = None
    fd_annual_rate: float | None = None
    fd_start_date: str | None = None
    mf_scheme_code: str | None = None
    mf_units_held: float | None = None


class GoalValuationResult(BaseModel):
    goal_id: str
    current_value: float | None
    error: str | None
